### PR TITLE
RNN stage + attention stage

### PR DESCRIPTION
**Pull request by daniel03c1**
RNN stage는 complexity까지 다 완료했는데 attention stage는 모델 구상만 하고 
수정할 사항이 없거나, 피드백 받은 이후에 complexity를 구현하겠습니다.

### CODE REVIEW

File: modules.py [hunk 1]
```diff
@@ -411,6 +506,126 @@ def conformer_block(inputs):
...
+
+            if not layer_norm_in_front:
+                x = LayerNormalization()(x)
+
+        # Multi Head Self Attention
+        attn = x
+        pos_encoding = pos_encoding(x.shape)(x)
+
+        if layer_norm_in_front:
+            attn = LayerNormalization()(attn)
```
Comment (ParkJinHyeock) [inline modules.py:579]: 이게 layer_norm은 무조건 있다고 가정하고, 이걸 앞에 놓을지 뒤에 놓을지를 결정하는 변수로 사용하신 것 같은데, 어떤 의미가 혹시 있나요? 이 부분만 보면 처음이랑 끝부분말고는  중간중간 layer norm위치는 같은데

File: modules.py [hunk 2]
```diff
@@ -411,6 +506,126 @@ def conformer_block(inputs):
...
+        # GLU
+        conv = x
+        if use_glu:
+            if layer_norm_in_front:
+                conv = LayerNormalization()(conv)
+            conv = Conv1D(filters=2*d_model, kernel_size=1)(conv)
+            conv_1, conv_2 = tf.split(conv, 2, axis=-1)
+            conv_2 = tf.keras.activations.sigmoid(conv_2)
+            conv = conv_1 * conv_2
+
```
Comment (ParkJinHyeock) [inline modules.py:602]: 그리고 만약 그런방식으로 하면 이부분에도 if not부분 추가해야할 것 같습니다!

File: modules.py (hunk 1 above)
Comment (daniel03c1) [reply modules.py:579]: layernorm이 앞에 있는게 conformer고 뒤에 있는게 transformer인데 두 가지를 다 표현할 수 있도록 하려고 넣었습니다.
성능 차이가 크게 날지는 잘 모르겠는데, 둘다 표현할 수 있도록 하기 위해서 구현했습니다.

다른점이 layer norm을 먼저 하게 되면 x + func(layernorm(x))이 되고,
반대로 layer norm을 나중에 하게 되면 layernorm(x + func(x))는 데요,
전자 방식은 skip connection에는 layernorm이 없이 여러 모듈을 지나면서 결과 값들이 계속 더해지기만 방식이고,
후자 방식은 skip connection과 더한 이후에 layernorm이 되기 때문에 skip connection이 있는 구간마다 
매번 layernorm을 하게 됩니다.

File: modules.py (hunk 2 above)
Comment (daniel03c1) [reply modules.py:602]: 저도 그 생각을 하긴 했는데 진짜 논문 보니까 attention + GLU가 아니라 GLU + depthwise를 통으로 보는거 같아서
GLU만 사용하면 일종의 attention의 activation으로 보고 (normalization 바로 뒤 activation),
depthwise만 사용하면, 그에 맞게 layer_norm_in_front를 적용하고,
둘다 사용하면 기존 conformer처럼 GLU + depthwise를 통으로 보고, 그 통의 앞뒤에만 layernorm을 적용하려고 했는데
혹시 말씀하신게 그게 맞을까요?

Comment (ParkJinHyeock) [reply modules.py:602]: 음 그런 의도라면 괜찮을 것 같습니다!

File: complexity.py [hunk 3]
```diff
@@ -356,22 +374,46 @@ def linear_complexity(input_shape, units, use_bias=True, prev_cx=None):
...
+def lstm_complexity(input_shape, units, use_bias=True,
+                   bi=True, merge_mode='mul', prev_cx=None):
+    num_steps, input_chan = input_shape[-2:]
+
+    params = 4 * units * (input_chan + units + use_bias)
+    if bi:
+        params *= 2
+
+    # https://github.com/Lyken17/pytorch-OpCounter/blob/master/thop/rnn_hooks.py
+    flops = num_steps * (units + input_chan + 2 * use_bias + 1) * units * 4
```
Comment (ParkJinHyeock) [inline complexity.py]: Complexity 부분은 문제 없는 것 같습니다!

**Review by ParkJinHyeock — APPROVED** [review]
크게 문제있는 부분은 없는 것 같습니다!
